Feature/nav component for child fragments

nbradbury opened the pull request:
Closes #1134 - This PR is a replacement for #1136. In that PR I had to hack around the nav component's limitations with bottom navigation fragments, and when I started adding child fragments (such as product detail) to the mix even more hacks were required.

In this PR the bottom navigation fragments remain as they were (ie: no nav component), but all the child fragments are part of the navigation graph. In order for this to work, I had to add an empty "root fragment" to the main activity because the nav component requires a non-activity start destination.

Note that when the code refers to a "child fragment" it really means a fragment that's part of the navigation graph (ie: not one of the three top level fragments). I'd be happy to change this but couldn't think of a better name :)

~~I'm submitting this as a draft for feedback. There are still quite a few things I need to address:~~
- [x]  Bottom nav doesn't re-appear between navigations
- [x]  Order list menu appears in order detail & other children
- [x]  Check if we still need to to override `onBackStackChanged` in top level fragments
- [x]  Check if `TopLevelFragment runOnResumeFunc` is still necessary
- [x]  Clean up any remaining `childFragmentManager` code (and related code)
- [x]  Top level fragments scrolls to the top when bottom nav reselected even when child fragment is showing
- [x]  Top level fragment flashes briefly before showing order fulfillment, add tracking, and add note
- [x]  Add transitions
- [x] Order list is empty when returning to it if search results were previously showing
- [x] Crash when returning to main activity from settings
- [x] Top level fragments are blank when opening settings from a child fragment then returning to the top level fragment
- [x]  Fix failing tests

I'm **much** more comfortable with this PR than I am with #1136. If/when the nav component better handles bottom navigation, we can look into moving the top level fragments to the navigation graph.

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

nbradbury commented:
> I'm not sure if you are finished with the Product Review child fragment but when I click on an item in the Notification tab, I got this screen:

My bad. I fixed that over the weekend but forgot to push the change. It's in 2fb92d8.

> I think the extra icons from OrderListFragment are still visible in child fragments though. 

Yep, that's the second checklist item above.

> The Order list screen was blank was me in one scenario but I wasn't able to reproduce it.

I'm not able to repro that either, but it doesn't surprise. There's a lot of logic that took place when the `childFragmentManager` backstack changed which isn't firing now. I may end up punting that to a separate PR because this one is already too large.

nbradbury commented:
> I'm not sure if you are finished with the Product Review child fragment but when I click on an item in the Notification tab, I got this screen:

My bad. I fixed that over the weekend but forgot to push the change. It's in 2fb92d8.

> I think the extra icons from OrderListFragment are still visible in child fragments though. 

Yep, that's the second checklist item above.

> The Order list screen was blank was me in one scenario but I wasn't able to reproduce it.

I'm not able to repro that either, but it doesn't surprise. There's a lot of logic that took place when the `childFragmentManager` backstack changed which isn't firing now. I may end up punting that to a separate PR because this one is already too large.

anitaa1990 commented:
Hey @nbradbury, I faced a couple of minor issues when testing.

The Order list fragment is blank after I change an Order Status from the `OrderDetailFragment`.
<img width="300" src="https://user-images.githubusercontent.com/22608780/59756468-639df200-92a7-11e9-8932-e6dcf6311a79.gif">

When I mark an unfulfilled order as complete, I am redirected to the Order Detail page. When I click on back from that screen, I see the Order Fullfil screen again. The order is marked as complete though in Order list screen.
<img width="300" src="https://user-images.githubusercontent.com/22608780/59756686-e0c96700-92a7-11e9-9a0b-d3732ef44f9c.gif">

I also faced the same crash that Amanda did but when clicking back from the Help screen. Again, happens only when `Do not keep activities` is enabled.
```
Attempt to invoke virtual method 'org.wordpress.android.fluxc.store.WCStatsStore$StatsGranularity com.woocommerce.android.ui.dashboard.DashboardStatsView.getActiveGranularity()' on a null object reference
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2659)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2724)
        at android.app.ActivityThread.-wrap12(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1473)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6123)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:867)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:757)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'org.wordpress.android.fluxc.store.WCStatsStore$StatsGranularity com.woocommerce.android.ui.dashboard.DashboardStatsView.getActiveGranularity()' on a null object reference
        at com.woocommerce.android.ui.dashboard.DashboardFragment.refreshDashboard(DashboardFragment.kt:252)
        at com.woocommerce.android.ui.dashboard.DashboardFragment.onReturnedFromChildFragment(DashboardFragment.kt:146)
        at com.woocommerce.android.ui.main.MainActivity.onDestinationChanged(MainActivity.kt:291)
        at androidx.navigation.NavController.addOnDestinationChangedListener(NavController.java:218)
        at com.woocommerce.android.ui.main.MainActivity.onCreate(MainActivity.kt:142)
        at android.app.Activity.performCreate(Activity.java:6672)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1140)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2612)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2724) 
        at android.app.ActivityThread.-wrap12(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1473) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:154) 
        at android.app.ActivityThread.main(ActivityThread.java:6123)
```

When `Do not keep activities` is enabled, when I click on the `Call` or `Message` button from an order and come back to the app and click on back, the fragment is blank. 
<img width="300" src="https://user-images.githubusercontent.com/22608780/59757273-0dca4980-92a9-11e9-9f4e-8e7ce29f00d7.gif">

nbradbury commented:
@AmandaRiu @anitaa1990 Thanks for all the testing! I've addressed all the issues so ready for another round 🤞 

> minimizing the app when on the "my store" tab and restoring causes a crash

This crash and all the others had the same root cause, which also caused the blank fragments. The problem was a navigation event was being fired when the activity was recreated - fixed in d45c970.

> The bottom navigation bar is visible when adding an order note or shipment tracking...Is there a way to hide the bottom nav when on these views?

Added in b80f4d7.

> Unable to moderate a product review - throws an error

Fixed in 8f1aa28.

> Opening a new order notification or product review shows the "mark all as read" option in the toolbar

Fixed in 4205c4f.

> When I mark an unfulfilled order as complete, I am redirected to the Order Detail page. When I click on back from that screen, I see the Order Fullfil screen again. 

Fixed in 5480676.

nbradbury commented:
> I am facing the blank fragment issue every time I go out of the app and come back again.

Bah, sorry - really thought that was fixed. It should be fixed now by 9564f53.

> the toolbar title does not change back to `Notifications`

Do you mind if I address that while working on #1172? I should have a PR for that ready to go right after this one is merged.

> When marking a review item as approve/unapprove, the changes are not reflected in the review detail screen

Hmmm, I'm able to repro this but not consistently. I'll ping you when I've figured it out.

nbradbury commented:
@anitaa1990 It turns out the moderation bug **does** happen in `develop` as well. It appears to be a timing issue - it happens if you moderate the review, then tap the review in the list *after* the snackbar disappears but *before* the request to change the status completes.

So, this PR is ready for another round.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -161,31 +176,96 @@ class MainActivity : AppCompatActivity(),
+    /**
+     * The current fragment in the nav controller has changed - make sure the actionbar's up icon is correct
+     */
+    override fun onDestinationChanged(controller: NavController, destination: NavDestination, arguments: Bundle?) {
```
Agreed! The only thing I think we can improve here (later, not a priority now) is maybe have an argument called "showBottomBar" with a value of "true|false" that could be passed around to all the destinations. Then in this method we'd only need to look for that argument to determine if the bottombar should be visible or not, instead of having a switch statement that looked at each destination individually. I'm just thinking in the future this would be easier to remember than the switch statement.

AmandaRiu replied:
...and building off of that, we could make it so that if `showBottomBar` is false, then we automatically set `showCrossIcon` to `true`.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -424,8 +560,19 @@ class MainActivity : AppCompatActivity(),
+            navController.popBackStack(R.id.orderDetailFragment, true)
+        }
+        val orderId = OrderIdentifier(localSiteId, remoteOrderId)
+        val action = OrderDetailFragmentDirections.actionGlobalOrderDetailFragment(orderId, remoteNoteId, markComplete)
```
I love these global actions! It makes composing a destination so concise.
